File: Backend/app/pw_library/rust_methods/bisection.rs
// Define a struct to store results for each iteration
#[derive(Debug)]
struct BisectionResult {
    iteration: usize,
    x_i: f64,
    f_x_i: f64,
    error: f64,
}

// Bisection method function in Rust
fn bisection_method<F>(
    f: F,
    mut a: f64,
    mut b: f64,
    tolerance: f64,
    max_iterations: usize,
) -> (f64, usize, bool, Vec<BisectionResult>)
where
    F: Fn(f64) -> f64,
{
    // Check if f(a) and f(b) have opposite signs
    if f(a) * f(b) >= 0.0 {
        panic!("f(a) and f(b) must have opposite signs.");
    }

    // Create a Vec to store the results for each iteration
    let mut result_array: Vec<BisectionResult> = Vec::new();

    let mut c = (a + b) / 2.0; // Midpoint initialization

    // Bisection loop
    for i in 0..max_iterations {
        c = (a + b) / 2.0; // Midpoint

        // Check if the solution has converged
        if (b - a).abs() < tolerance || f(c).abs() < tolerance {
            return (c, i + 1, true, result_array); // Converged
        }

        // Narrow the interval based on the sign of f(c)
        if f(a) * f(c) < 0.0 {
            b = c; // The root is in [a, c]
        } else {
            a = c; // The root is in [c, b]
        }

        // Store the result of the current iteration
        let result = BisectionResult {
            iteration: i + 1,
            x_i: c,
            f_x_i: f(c),
            error: (b - a).abs(),
        };
        result_array.push(result);
    }

    // If the loop finishes without converging, return the last midpoint and results
    (c, max_iterations, false, result_array) // Did not converge within max_iterations
}

fn main() {
    // Example usage of the bisection method
    let f = |x: f64| x.powi(3) - x - 2.0; // Example function f(x) = x^3 - x - 2

    let a = 1.0; // Left endpoint of the interval
    let b = 2.0; // Right endpoint of the interval
    let tolerance = 1e-7; // Tolerance for stopping criterion
    let max_iterations = 1000; // Maximum number of iterations

    // Call the bisection method
    let (root, iterations, converged, results) = bisection_method(f, a, b, tolerance, max_iterations);

    // Print the results
    println!(
        "Root: {}, Iterations: {}, Converged: {}",
        root, iterations, converged
    );

    // Print all results
    println!("Results from each iteration:");
    for result in results {
        println!(
            "Iteration: {}, x_i: {}, f_x_i: {}, Error: {}",
            result.iteration, result.x_i, result.f_x_i, result.error
        );
    }
}

File: Backend/app/pw_library/python_methods/bisection.py

import pandas as pd

from app.helpers.function_parser import string_function_evaluator


def bisection_method(f, a, b, tolerance=1e-7, max_iterations=100):
    """
    Bisection method to find the root of f(x) in the interval [a, b].

    Parameters:
    f : function
        The function for which we are trying to find a root.
    a, b : float
        The interval [a, b] in which the root is located.
    tolerance : float, optional
        The stopping criterion for the algorithm.
    max_iterations : int, optional
        The maximum number of iterations to perform.

    Returns:
    c : float
        The approximate root.
    iterations : int
        The number of iterations performed.
    converged : bool
        Whether the algorithm converged or not.
    """
    f_a = string_function_evaluator(f, a)
    f_b = string_function_evaluator(f, b)
    if (f_a * f_b) >= 0:
        raise ValueError("f(a) and f(b) must have opposite signs.")
    c = (a + b) / 2  # Midpoint initial
    result_array = []
    for i in range(max_iterations):
        c = (a + b) / 2  # Midpoint
        f_c = string_function_evaluator(f, c)
        if abs(b - a) < tolerance or abs(f_c) < tolerance:
            df_result = pd.DataFrame(data=result_array)
            return c, i + 1, True, df_result  # Converged

        # Narrow the interval based on the sign of f(c)
        if f_a * f_c < 0:
            b = c  # The root is in [a, c]
        else:
            a = c  # The root is in [c, b]

        result = {'i': i, 'x_i': c, 'f_x_i': f_c, 'e': abs(b - a)}
        result_array.append(result)

    return c, max_iterations, False  # Did not converge within max_iterations